[BUG] test_hash_* failed "java.util.NoSuchElementException: head of empty list" or "Too many times of repartition, may hit a bug?"

**Describe the bug**
this was found in different nightly pipelines with different spark versions, 3.2.0,3.3.4,3.4.0,3.5.3 etc

and the result is non-deterministic, ~30% runs repro (potentially with specific data_seeds)

```
src.main.python.hash_aggregate_test.test_hash_grpby_sum_count_action[true-1-None-[('a', Long), ('b', Integer), ('c', Long)]][DATAGEN_SEED=1732707822, TZ=UTC, INJECT_OOM]
src.main.python.hash_aggregate_test.test_hash_grpby_sum_count_action[true-1-1-[('a', Long), ('b', Integer), ('c', Long)]][DATAGEN_SEED=1732707822, TZ=UTC, INJECT_OOM]
src.main.python.hash_aggregate_test.test_hash_grpby_sum_count_action[false-1-None-[('a', Long), ('b', Integer), ('c', Long)]][DATAGEN_SEED=1732707822, TZ=UTC]
src.main.python.hash_aggregate_test.test_hash_grpby_sum_count_action[false-1-1-[('a', Long), ('b', Integer), ('c', Long)]][DATAGEN_SEED=1732707822, TZ=UTC, INJECT_OOM]

# some other failed data_seed we saw
[DATAGEN_SEED=1732707784, TZ=UTC]
[DATAGEN_SEED=1732734558, TZ=UTC]
[DATAGEN_SEED=1732742834, TZ=Canada/Newfoundland]
```

```
data_gen = [('a', Long), ('b', Integer), ('c', Long)]
override_split_until_size = None, override_batch_size_bytes = 1
kudo_enabled = 'false'

    @disable_ansi_mode  # https://github.com/NVIDIA/spark-rapids/issues/5114
    @pytest.mark.parametrize('data_gen', [_longs_with_nulls], ids=idfn)
    @pytest.mark.parametrize('override_split_until_size', [None, 1], ids=idfn)
    @pytest.mark.parametrize('override_batch_size_bytes', [None, 1], ids=idfn)
    @pytest.mark.parametrize("kudo_enabled", ["true", "false"], ids=idfn)
    def test_hash_grpby_sum_count_action(data_gen, override_split_until_size,
                                         override_batch_size_bytes, kudo_enabled):
        conf = {
            'spark.rapids.sql.test.overrides.splitUntilSize': override_split_until_size,
            kudo_enabled_conf_key: kudo_enabled
        }
        if override_batch_size_bytes is not None:
            conf["spark.rapids.sql.batchSizeBytes"] = override_batch_size_bytes
    
>       assert_gpu_and_cpu_row_counts_equal(
            lambda spark: gen_df(spark, data_gen, length=100).groupby('a').agg(f.sum('b')),
            conf = conf
        )

../../src/main/python/hash_aggregate_test.py:342: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../src/main/python/asserts.py:615: in assert_gpu_and_cpu_row_counts_equal
    _assert_gpu_and_cpu_are_equal(func, 'COUNT', conf=conf, is_cpu_first=is_cpu_first)
../../src/main/python/asserts.py:507: in _assert_gpu_and_cpu_are_equal
    from_gpu = run_on_gpu()
../../src/main/python/asserts.py:500: in run_on_gpu
    from_gpu = with_gpu_session(bring_back, conf=conf)
../../src/main/python/spark_session.py:166: in with_gpu_session
    return with_spark_session(func, conf=copy)
/opt/conda/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
../../src/main/python/spark_session.py:133: in with_spark_session
    ret = func(_spark)
../../src/main/python/asserts.py:212: in <lambda>
    bring_back = lambda spark: limit_func(spark).count()
../../../spark-3.2.0-bin-hadoop3.2/python/pyspark/sql/dataframe.py:680: in count
    return int(self._jdf.count())
/home/jenkins/agent/workspace/jenkins-rapids_it-AQE-pre_release-github-350/jars/spark-3.2.0-bin-hadoop3.2/python/lib/py4j-0.10.9.2-src.zip/py4j/java_gateway.py:1309: in __call__
    return_value = get_return_value(
../../../spark-3.2.0-bin-hadoop3.2/python/pyspark/sql/utils.py:111: in deco
    return f(*a, **kw)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

answer = 'xro116761'
gateway_client = <py4j.clientserver.JavaClient object at 0x7fa91867bfa0>
target_id = 'o116760', name = 'count'

    def get_return_value(answer, gateway_client, target_id=None, name=None):
        """Converts an answer received from the Java gateway into a Python object.
    
        For example, string representation of integers are converted to Python
        integer, string representation of objects are converted to JavaObject
        instances, etc.
    
        :param answer: the string returned by the Java gateway
        :param gateway_client: the gateway client used to communicate with the Java
            Gateway. Only necessary if the answer is a reference (e.g., object,
            list, map)
        :param target_id: the name of the object from which the answer comes from
            (e.g., *object1* in `object1.hello()`). Optional.
        :param name: the name of the member from which the answer comes from
            (e.g., *hello* in `object1.hello()`). Optional.
        """
        if is_error(answer)[0]:
            if len(answer) > 1:
                type = answer[1]
                value = OUTPUT_CONVERTER[type](answer[2:], gateway_client)
                if answer[1] == REFERENCE_TYPE:
>                   raise Py4JJavaError(
                        "An error occurred while calling {0}{1}{2}.\n".
                        format(target_id, ".", name), value)
E                   py4j.protocol.Py4JJavaError: An error occurred while calling o116760.count.
E                   : org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 3461.0 failed 1 times, most recent failure: Lost task 1.0 in stage 3461.0 (TID 11223) (100.96.98.224 executor 0): java.util.NoSuchElementException: head of empty list
E                   	at scala.collection.immutable.Nil$.head(List.scala:469)
E                   	at scala.collection.immutable.Nil$.head(List.scala:466)
E                   	at com.nvidia.spark.rapids.GpuAggregateIterator$.$anonfun$concatenateBatches$4(GpuAggregateExec.scala:710)
E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:48)
E                   	at com.nvidia.spark.rapids.GpuAggregateIterator$.$anonfun$concatenateBatches$2(GpuAggregateExec.scala:709)
E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
E                   	at com.nvidia.spark.rapids.GpuAggregateIterator$.$anonfun$concatenateBatches$1(GpuAggregateExec.scala:707)
E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$AutoCloseableAttemptSpliterator.next(RmmRapidsRetryIterator.scala:513)
E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.next(RmmRapidsRetryIterator.scala:649)
E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryAutoCloseableIterator.next(RmmRapidsRetryIterator.scala:553)
E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.drainSingleWithVerification(RmmRapidsRetryIterator.scala:291)
E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.withRetryNoSplit(RmmRapidsRetryIterator.scala:164)
E                   	at com.nvidia.spark.rapids.GpuAggregateIterator$.concatenateBatches(GpuAggregateExec.scala:702)
E                   	at com.nvidia.spark.rapids.AggregateUtils$.$anonfun$concatenateAndMerge$1(GpuAggregateExec.scala:145)
E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:66)
E                   	at com.nvidia.spark.rapids.AggregateUtils$.concatenateAndMerge(GpuAggregateExec.scala:144)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$RepartitionAggregateIterator.$anonfun$next$13(GpuAggregateExec.scala:1086)
E                   	at com.nvidia.spark.rapids.Arm$.closeOnExcept(Arm.scala:126)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$RepartitionAggregateIterator.$anonfun$next$12(GpuAggregateExec.scala:1076)
E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$RepartitionAggregateIterator.next(GpuAggregateExec.scala:1069)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$RepartitionAggregateIterator.next(GpuAggregateExec.scala:1060)
E                   	at com.nvidia.spark.rapids.CloseableBufferedIterator.next(CloseableBufferedIterator.scala:65)
E                   	at com.nvidia.spark.rapids.CloseableBufferedIterator.head(CloseableBufferedIterator.scala:47)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$ConcatIterator.$anonfun$next$10(GpuAggregateExec.scala:1046)
E                   	at com.nvidia.spark.rapids.Arm$.closeOnExcept(Arm.scala:126)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$ConcatIterator.next(GpuAggregateExec.scala:1044)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$ConcatIterator.next(GpuAggregateExec.scala:1033)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator.$anonfun$next$2(GpuAggregateExec.scala:946)
E                   	at scala.Option.map(Option.scala:230)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator.next(GpuAggregateExec.scala:946)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator.next(GpuAggregateExec.scala:898)
E                   	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
E                   	at com.nvidia.spark.rapids.DynamicGpuPartialAggregateIterator.$anonfun$next$16(GpuAggregateExec.scala:2199)
E                   	at scala.Option.map(Option.scala:230)
E                   	at com.nvidia.spark.rapids.DynamicGpuPartialAggregateIterator.next(GpuAggregateExec.scala:2199)
E                   	at com.nvidia.spark.rapids.DynamicGpuPartialAggregateIterator.next(GpuAggregateExec.scala:2047)
E                   	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase$$anon$1.partNextBatch(GpuShuffleExchangeExecBase.scala:335)
E                   	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase$$anon$1.hasNext(GpuShuffleExchangeExecBase.scala:357)
E                   	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:168)
E                   	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
E                   	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
E                   	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
E                   	at org.apache.spark.scheduler.Task.run(Task.scala:131)
E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1462)
E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
E                   	at java.lang.Thread.run(Thread.java:750)
E                   
E                   Driver stacktrace:
E                   	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2403)
E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2352)
E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2351)
E                   	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
E                   	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
E                   	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
E                   	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2351)
E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1109)
E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1109)
E                   	at scala.Option.foreach(Option.scala:407)
E                   	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1109)
E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2591)
E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2533)
E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2522)
E                   	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
E                   	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:898)
E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2214)
E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2235)
E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2254)
E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2279)
E                   	at org.apache.spark.rdd.RDD.$anonfun$collect$1(RDD.scala:1030)
E                   	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
E                   	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
E                   	at org.apache.spark.rdd.RDD.withScope(RDD.scala:414)
E                   	at org.apache.spark.rdd.RDD.collect(RDD.scala:1029)
E                   	at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:394)
E                   	at org.apache.spark.sql.Dataset.$anonfun$count$1(Dataset.scala:3012)
E                   	at org.apache.spark.sql.Dataset.$anonfun$count$1$adapted(Dataset.scala:3011)
E                   	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3706)
E                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
E                   	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
E                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
E                   	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775)
E                   	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
E                   	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3704)
E                   	at org.apache.spark.sql.Dataset.count(Dataset.scala:3011)
E                   	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
E                   	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
E                   	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
E                   	at java.lang.reflect.Method.invoke(Method.java:498)
E                   	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
E                   	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
E                   	at py4j.Gateway.invoke(Gateway.java:282)
E                   	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
E                   	at py4j.commands.CallCommand.execute(CallCommand.java:79)
E                   	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
E                   	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
E                   	at java.lang.Thread.run(Thread.java:750)
E                   Caused by: java.util.NoSuchElementException: head of empty list
E                   	at scala.collection.immutable.Nil$.head(List.scala:469)
E                   	at scala.collection.immutable.Nil$.head(List.scala:466)
E                   	at com.nvidia.spark.rapids.GpuAggregateIterator$.$anonfun$concatenateBatches$4(GpuAggregateExec.scala:710)
E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:48)
E                   	at com.nvidia.spark.rapids.GpuAggregateIterator$.$anonfun$concatenateBatches$2(GpuAggregateExec.scala:709)
E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
E                   	at com.nvidia.spark.rapids.GpuAggregateIterator$.$anonfun$concatenateBatches$1(GpuAggregateExec.scala:707)
E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$AutoCloseableAttemptSpliterator.next(RmmRapidsRetryIterator.scala:513)
E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.next(RmmRapidsRetryIterator.scala:649)
E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryAutoCloseableIterator.next(RmmRapidsRetryIterator.scala:553)
E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.drainSingleWithVerification(RmmRapidsRetryIterator.scala:291)
E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.withRetryNoSplit(RmmRapidsRetryIterator.scala:164)
E                   	at com.nvidia.spark.rapids.GpuAggregateIterator$.concatenateBatches(GpuAggregateExec.scala:702)
E                   	at com.nvidia.spark.rapids.AggregateUtils$.$anonfun$concatenateAndMerge$1(GpuAggregateExec.scala:145)
E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:66)
E                   	at com.nvidia.spark.rapids.AggregateUtils$.concatenateAndMerge(GpuAggregateExec.scala:144)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$RepartitionAggregateIterator.$anonfun$next$13(GpuAggregateExec.scala:1086)
E                   	at com.nvidia.spark.rapids.Arm$.closeOnExcept(Arm.scala:126)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$RepartitionAggregateIterator.$anonfun$next$12(GpuAggregateExec.scala:1076)
E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$RepartitionAggregateIterator.next(GpuAggregateExec.scala:1069)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$RepartitionAggregateIterator.next(GpuAggregateExec.scala:1060)
E                   	at com.nvidia.spark.rapids.CloseableBufferedIterator.next(CloseableBufferedIterator.scala:65)
E                   	at com.nvidia.spark.rapids.CloseableBufferedIterator.head(CloseableBufferedIterator.scala:47)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$ConcatIterator.$anonfun$next$10(GpuAggregateExec.scala:1046)
E                   	at com.nvidia.spark.rapids.Arm$.closeOnExcept(Arm.scala:126)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$ConcatIterator.next(GpuAggregateExec.scala:1044)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator$ConcatIterator.next(GpuAggregateExec.scala:1033)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator.$anonfun$next$2(GpuAggregateExec.scala:946)
E                   	at scala.Option.map(Option.scala:230)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator.next(GpuAggregateExec.scala:946)
E                   	at com.nvidia.spark.rapids.GpuMergeAggregateIterator.next(GpuAggregateExec.scala:898)
E                   	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
E                   	at com.nvidia.spark.rapids.DynamicGpuPartialAggregateIterator.$anonfun$next$16(GpuAggregateExec.scala:2199)
E                   	at scala.Option.map(Option.scala:230)
E                   	at com.nvidia.spark.rapids.DynamicGpuPartialAggregateIterator.next(GpuAggregateExec.scala:2199)
E                   	at com.nvidia.spark.rapids.DynamicGpuPartialAggregateIterator.next(GpuAggregateExec.scala:2047)
E                   	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase$$anon$1.partNextBatch(GpuShuffleExchangeExecBase.scala:335)
E                   	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase$$anon$1.hasNext(GpuShuffleExchangeExecBase.scala:357)
E                   	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:168)
E                   	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
E                   	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
E                   	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
E                   	at org.apache.spark.scheduler.Task.run(Task.scala:131)
E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1462)
E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
E                   	... 1 more
```



**Steps/Code to reproduce bug**
Please provide a list of steps or a code sample to reproduce the issue.
Avoid posting private or sensitive data.

**Expected behavior**
A clear and concise description of what you expected to happen.

**Environment details (please complete the following information)**
 - Environment location: [Standalone, YARN, Kubernetes, Cloud(specify cloud provider)]
 - Spark configuration settings related to the issue

**Additional context**
Add any other context about the problem here.

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuAggregateExec.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuAggregateExec.scala
@@ -219,9 +219,6 @@ object AggregateUtils extends Logging {
   ): Boolean = {
 
     var repartitionHappened = false
-    if (hashSeed > 200) {
-      throw new IllegalStateException("Too many times of repartition, may hit a bug?")
-    }
 
     def repartitionAndClose(batch: SpillableColumnarBatch): Unit = {
 
@@ -280,15 +277,23 @@ object AggregateUtils extends Logging {
 
       val newBuckets = batchesByBucket.flatMap(bucket => {
         if (needRepartitionAgain(bucket)) {
-          val nextLayerBuckets =
-            ArrayBuffer.fill(hashBucketNum)(new AutoClosableArrayBuffer[SpillableColumnarBatch]())
-          // Recursively merge and repartition the over sized bucket
-          repartitionHappened =
-            iterateAndRepartition(
-              new CloseableBufferedIterator(bucket.iterator), metrics, targetMergeBatchSize,
-              helper, hashKeys, hashBucketNum, hashSeed + 7,
-              nextLayerBuckets) || repartitionHappened
-          nextLayerBuckets
+          if (hashSeed + 7 > 200) {
+            log.warn("Too many times of repartition, may hit a bug? Size for each batch in " +
+              "current bucket: " + bucket.map(_.sizeInBytes).mkString(", ") + " rows: " +
+              bucket.map(_.numRows()).mkString(", ") + " targetMergeBatchSize: "
+              + targetMergeBatchSize)
+            ArrayBuffer.apply(bucket)
+          } else {
+            val nextLayerBuckets =
+              ArrayBuffer.fill(hashBucketNum)(new AutoClosableArrayBuffer[SpillableColumnarBatch]())
+            // Recursively merge and repartition the over sized bucket
+            repartitionHappened =
+              iterateAndRepartition(
+                new CloseableBufferedIterator(bucket.iterator), metrics, targetMergeBatchSize,
+                helper, hashKeys, hashBucketNum, hashSeed + 7,
+                nextLayerBuckets) || repartitionHappened
+            nextLayerBuckets
+          }
         } else {
           ArrayBuffer.apply(bucket)
         }
@@ -1075,8 +1080,8 @@ class GpuMergeAggregateIterator(
           closeOnExcept(new ArrayBuffer[AutoClosableArrayBuffer[SpillableColumnarBatch]]) {
             toAggregateBuckets =>
               var currentSize = 0L
-              while (batchesByBucket.nonEmpty &&
-                batchesByBucket.last.size() + currentSize < targetMergeBatchSize) {
+              while (batchesByBucket.nonEmpty && (toAggregateBuckets.isEmpty ||
+                batchesByBucket.last.size() + currentSize < targetMergeBatchSize)) {
                 val bucket = batchesByBucket.remove(batchesByBucket.size - 1)
                 currentSize += bucket.map(_.sizeInBytes).sum
                 toAggregateBuckets += bucket

